src/java.base/share/classes/sun/security/ssl/KAKeyDerivation.java

```diff
@@ -86,45 +87,59 @@ private SecretKey t12DeriveKey(String algorithm,
             }
             SSLKeyDerivation kd = mskd.createKeyDerivation(
                     context, preMasterSecret);
-            return kd.deriveKey("MasterSecret", params);
+            return kd.deriveKey("MasterSecret");
         } catch (GeneralSecurityException gse) {
             throw new SSLHandshakeException("Could not generate secret", gse);
+        } finally {
+            if (preMasterSecret instanceof SecretKeySpec s) {
+                SharedSecrets.getJavaxCryptoSpecAccess().clearSecretKeySpec(s);
+            }
         }
     }
 
     /**
      * Handle the TLSv1.3 objects, which use the HKDF algorithms.
      */
-    private SecretKey t13DeriveKey(String algorithm,
-            AlgorithmParameterSpec params) throws IOException {
+    private SecretKey t13DeriveKey(String type)
+            throws IOException {
+        SecretKey sharedSecret = null;
+        SecretKey earlySecret = null;
+        SecretKey saltSecret = null;
         try {
             KeyAgreement ka = KeyAgreement.getInstance(algorithmName);
             ka.init(localPrivateKey);
             ka.doPhase(peerPublicKey, true);
-            SecretKey sharedSecret
-                    = ka.generateSecret("TlsPremasterSecret");
+            sharedSecret = ka.generateSecret("TlsPremasterSecret");
 
             CipherSuite.HashAlg hashAlg = context.negotiatedCipherSuite.hashAlg;
             SSLKeyDerivation kd = context.handshakeKeyDerivation;
-            HKDF hkdf = new HKDF(hashAlg.name);
+            KDF hkdf = KDF.getInstance(hashAlg.hkdfAlgorithm);
             if (kd == null) {   // No PSK is in use.
-                // If PSK is not in use Early Secret will still be
+                // If PSK is not in use, Early Secret will still be
                 // HKDF-Extract(0, 0).
                 byte[] zeros = new byte[hashAlg.hashLength];
-                SecretKeySpec ikm
-                        = new SecretKeySpec(zeros, "TlsPreSharedSecret");
-                SecretKey earlySecret
-                        = hkdf.extract(zeros, ikm, "TlsEarlySecret");
+                earlySecret = hkdf.deriveKey("TlsEarlySecret",
+                        HKDFParameterSpec.ofExtract().addSalt(zeros)
+                        .addIKM(zeros).extractOnly());
                 kd = new SSLSecretDerivation(context, earlySecret);
             }
 
             // derive salt secret
-            SecretKey saltSecret = kd.deriveKey("TlsSaltSecret", null);
+            saltSecret = kd.deriveKey("TlsSaltSecret");
 
             // derive handshake secret
-            return hkdf.extract(saltSecret, sharedSecret, algorithm);
+            return hkdf.deriveKey(type, HKDFParameterSpec.ofExtract()
```

Comment:
The line above may fail because the `hkdf` object has been used once on line 121 with zero-valued salt and IKM, which selected the software-based HKDF from SunJCE. At this point, however, `sharedSecret` is the result of ECDH and may be non-extractable if produced by an HSM, making it incompatible with the SunJCE implementation. To avoid this issue, get a new `hkdf` by calling `hkdf = KDF.getInstance(hashAlg.hkdfAlgorithm)` before this line.

Reply:
Hmm, ok, interesting scenario. I add another `KDF.getInstance()` call as you suggested just to be safe.

Reply:
I think that deserves a code comment; it's far from obvious why we do that. Also, we will make P11 work with zero-valued salt soon.